keyvault/client/keys/src/main/java/com/azure/keyvault/keys/KeyAsyncClient.java

```diff
@@ -0,0 +1,593 @@
+     * {@link KeyAsyncClient#getKey(KeyBase baseKey)} within {@link Flux#flatMap(Function)}.</p>
+     * <pre>
+     * Flux&lt;Key&gt; keys = keyAsyncClient.listKeyVersions("keyName")
+     *   .flatMap(keyAsyncClient::getKey)
+     *   .map(Response::value);
+     * </pre>
+     *
+     * @param name The name of the key.
+     * @throws ResourceNotFoundException when a key with {@code name} doesn't exist in the key vault.
+     * @throws HttpRequestException when a key with {@code name} is empty string.
+     * @return A {@link Flux} containing {@link KeyBase key} of all the versions of the specified key in the vault. Flux is empty if key with {@code name} does not exist in key vault.
+     */
+    public Flux<KeyBase> listKeyVersions(String name) {
+        return service.getKeyVersions(endpoint, name, DEFAULT_MAX_PAGE_RESULTS, API_VERSION, ACCEPT_LANGUAGE, CONTENT_TYPE_HEADER_VALUE).flatMapMany(this::extractAndFetchKeys);
+    }
+
+    /**
+     * Gets attributes of all the keys given by the {@code nextPageLink} that was retrieved from a call to
+     * {@link KeyAsyncClient#listKeys()}.
+     *
+     * @param nextPageLink The {@link KeyBasePage#nextLink()} from a previous, successful call to one of the list operations.
+     * @return A stream of {@link KeyBase key} from the next page of results.
+     */
+    private Flux<KeyBase> listKeysNext(String nextPageLink) {
+        return service.getKeys(endpoint, nextPageLink, ACCEPT_LANGUAGE, CONTENT_TYPE_HEADER_VALUE).flatMapMany(this::extractAndFetchKeys);
+    }
+
+    private Publisher<KeyBase> extractAndFetchKeys(PagedResponse<KeyBase> page) {
+        return extractAndFetch(page, this::listKeysNext);
+    }
+
+    /**
+     * Gets attributes of all the keys given by the {@code nextPageLink} that was retrieved from a call to
+     * {@link KeyAsyncClient#listDeletedKeys()}.
+     *
+     * @param nextPageLink The {@link DeletedKeyPage#nextLink()} from a previous, successful call to one of the list operations.
+     * @return A stream of {@link KeyBase key} from the next page of results.
+     */
+    private Flux<DeletedKey> listDeletedKeysNext(String nextPageLink) {
+        return service.getDeletedKeys(endpoint, nextPageLink, ACCEPT_LANGUAGE, CONTENT_TYPE_HEADER_VALUE).flatMapMany(this::extractAndFetchDeletedKeys);
+    }
+
+    private Publisher<DeletedKey> extractAndFetchDeletedKeys(PagedResponse<DeletedKey> page) {
+        return extractAndFetch(page, this::listDeletedKeysNext);
+    }
+
+    //TODO: Extract this in azure-core ImplUtils and use from there
+    private <T> Publisher<T> extractAndFetch(PagedResponse<T> page, Function<String, Publisher<T>> content) {
+        String nextPageLink = page.nextLink();
+        if (nextPageLink == null) {
+            return Flux.fromIterable(page.items());
+        }
+        return Flux.fromIterable(page.items()).concatWith(content.apply(nextPageLink));
+    }
```

Comment:
This exists in core now, you should get rid of this method.

Reply:
Yeah JsonWebKey won't be renamed.
@schaabs, but can we rename JsonWebKeyType to KeyType, JsonWebKeyOperation to KeyOperation etc.?